consensus/ibft/operator_service.go

```diff
@@ -2,163 +2,234 @@ package ibft
+	}
 
-		count := snap.Count(func(v *Vote) bool {
-			return v.Address == addr && v.Validator == o.ibft.validatorKeyAddr
-		})
+	resp := &proto.Snapshot{
+		Number:     height,
+		Hash:       header.Hash.String(),
+		Validators: validatorsToProtoValidators(validators),
+	}
 
-		if count == 0 {
-			// Candidate found
-			candidate = c
+	votes, err := getVotes(validatorsStore, height)
+	if err != nil {
+		return nil, err
+	}
 
-			break
-		}
+	if votes == nil {
+		// current ValidatorStore doesn't have voting function
+		return resp, nil
 	}
 
-	return candidate
+	resp.Votes = votesToProtoVotes(votes)
+
+	return resp, nil
 }
 
-// GetSnapshot returns the snapshot, based on the passed in request
-func (o *operator) GetSnapshot(ctx context.Context, req *proto.SnapshotReq) (*proto.Snapshot, error) {
-	var snap *Snapshot
+// Propose proposes a new candidate to be added / removed from the validator set
+func (o *operator) Propose(ctx context.Context, req *proto.Candidate) (*empty.Empty, error) {
+	votableSet, err := o.getVotableValidatorStore()
+	if err != nil {
+		return nil, err
+	}
 
-	var err error
+	candidate, err := o.parseCandidate(req)
+	if err != nil {
+		return nil, err
+	}
 
-	if req.Latest {
-		snap, err = o.ibft.getLatestSnapshot()
-	} else {
-		snap = o.ibft.getSnapshot(req.Number)
-		if snap == nil {
-			err = errSnapshotNotFound
-		}
+	if err := votableSet.Propose(candidate, req.Auth, o.ibft.currentSigner.Address()); err != nil {
+		return nil, err
 	}
 
+	return &empty.Empty{}, nil
+}
+
+// Candidates returns the validator candidates list
+func (o *operator) Candidates(ctx context.Context, req *empty.Empty) (*proto.CandidatesResp, error) {
+	votableValSet, err := o.getVotableValidatorStore()
 	if err != nil {
 		return nil, err
 	}
 
-	resp := snap.ToProto()
+	candidates := votableValSet.Candidates()
 
-	return resp, nil
+	return &proto.CandidatesResp{
+		Candidates: candidatesToProtoCandidates(candidates),
+	}, nil
 }
 
-// Propose proposes a new candidate to be added / removed from the validator set
-func (o *operator) Propose(ctx context.Context, req *proto.Candidate) (*empty.Empty, error) {
-	var addr types.Address
-	if err := addr.UnmarshalText([]byte(req.Address)); err != nil {
+// parseCandidate parses proto.Candidate and maps to validator
+func (o *operator) parseCandidate(req *proto.Candidate) (validators.Validator, error) {
+	signer, err := o.getLatestSigner()
+	if err != nil {
 		return nil, err
 	}
 
-	// check if the candidate is already there
-	o.candidatesLock.Lock()
-	defer o.candidatesLock.Unlock()
-
-	for _, c := range o.candidates {
-		if c.Address == req.Address {
-			return nil, fmt.Errorf("already a candidate")
+	switch signer.Type() {
+	case validators.ECDSAValidatorType:
+		return &validators.ECDSAValidator{
+			Address: types.StringToAddress(req.Address),
+		}, nil
+
+	case validators.BLSValidatorType:
+		// safe check
+		if req.Auth {
+			if _, err := crypto.UnmarshalBLSPublicKey(req.BlsPubkey); err != nil {
```

Comment:
I would maybe add a check (probably with some better description
```
If req.BlsPubkey == nil {
    return nil, errors.New("Missing BLS public key")
}
```
I think this would be useful for clients because --bls flag is not required. And if user doesn't add it whili proposing in BLS mode he would get an error 

`Public key must be 48 bytes `

which does not indicate that the bls key is missing

What do you think :)

Reply:
Fixed 5a4b11b1f137a81a76eb6d61ac7843b4d574d8c4